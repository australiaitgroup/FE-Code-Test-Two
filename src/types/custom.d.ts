// Defer the types, refer to https://webpack.js.org/guides/typescript/#importing-other-assets

declare module '*.png' {
    const content: any;
    export default content;
}

declare module '*.svg' {
    const content: any;
    export default content;
}
