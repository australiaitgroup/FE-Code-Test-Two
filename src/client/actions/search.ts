import fetch from '../utils/fetch';
import generateURL from '../utils/searchUrlHelpers';
import { ISearchResult } from '../../types/search';

// action types

export const SEARCH = 'SEARCH';
export const HANDLE_SEARCH_RESULT = 'HANDLE_SEARCH_RESULT';
export const HANDLE_INPUT_CHANGE = 'HANDLE_INPUT_CHANGE';
export const HANDLE_PAGE_CHANGE = 'HANDLE_PAGE_CHANGE';
export const SET_SEARCH_ERROR = 'SET_SEARCH_ERROR';

// action creators
â€‹
export function searchAction(searchFor: string) {
    return { searchFor, type: SEARCH };
}

export function handleSearchResultAction(result: ISearchResult) {
    return { result, type: HANDLE_SEARCH_RESULT };
}

export function handleInputChangeAction(input: string) {
    return { input, type: HANDLE_INPUT_CHANGE };
}

export function handlePageChangeAction(page: number) {
    return { page, type: HANDLE_PAGE_CHANGE };
}

export function setSearchErrorAction(error: Error) {
    return { error, type: SET_SEARCH_ERROR };
}

// async actions

export function doSearch(page?: number) {
    return (dispatch, getState) => {
        const input = getState().search.input;
        const settings = getState().settings;
        const url = generateURL(settings, input, page);

        if (page) {
            dispatch(handlePageChangeAction(page));
        } else {
            dispatch(searchAction(input));
        }

        return fetch(url)
            .then(result => dispatch(handleSearchResultAction(result)))
            .catch(err => dispatch(setSearchErrorAction(err)));
    };
}
