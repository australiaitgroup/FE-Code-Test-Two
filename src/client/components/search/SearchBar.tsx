import * as React from 'react';
import { Button, Input } from 'semantic-ui-react';

import { ISearchBarProps } from '../interfaces/searchBar.interface';

let searchTimer;

const SearchBar: React.FunctionComponent<ISearchBarProps> = ({
    handleInputChange,
    search,
}) => (
    <Input
        action={
            <Button
                color="teal"
                content="Search"
                onClick={search}
            />
        }
        fluid
        icon="search"
        iconPosition="left"
        onKeyUp={(event) => {
            clearTimeout(searchTimer);
            handleInputChange(event.target.value);
            if (event.key === 'Enter') return search();
            searchTimer = setTimeout(search, 300);
        }}
        placeholder="Search..."
    />
);

export default SearchBar;
