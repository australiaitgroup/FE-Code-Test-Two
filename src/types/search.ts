export interface ISearchResult {
    hits: IHit[];
    nbHits: number;
    page: number;
    nbPages: number;
    hitsPerPage: number;
    processingTimeMS: number;
    exhaustiveNbHits: boolean;
    query: string;
    params: string;
}

export interface IHit {
    created_at: string;
    title: string | null;
    url: string | null;
    author: string;
    points: number;
    story_text: string | null;
    comment_text: string | null;
    num_comments: number | null;
    story_id: number | null;
    story_title: string | null;
    story_url: string | null;
    parent_id: number | null;
    created_at_i: number;
    _tags: string[];
    objectID: string;
    _highlightResult: {
        author: IHightlightResultItem;
        comment_text: IHightlightResultItem;
        story_text: IHightlightResultItem;
        story_title: IHightlightResultItem,
        title: IHightlightResultItem;
        url: IHightlightResultItem;
    };
}

interface IHightlightResultItem {
    value: string;
    matchLevel: string;
    fullyHighlighted?: boolean;
    matchedWords: string[];
}

export const enum SearchType {
    All = 'All',
    Stories = 'Stories',
    Comments = 'Comments',
}

export const enum SortOrder {
    Popularity = 'Popularity',
    Date = 'Date',
}

export const enum DateRange {
    AllTime = 'All Time',
    LastDay = 'Last 24 hours',
    PastWeek = 'Past Week',
    PastMonth = 'Past Month',
    PastYear = 'Past Year',
    Custom = 'Custom',
}
