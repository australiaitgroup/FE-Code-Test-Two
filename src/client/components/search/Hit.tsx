import * as moment from 'moment';
import * as React from 'react';
import MediaQuery from 'react-responsive';
import { Icon, Item, Label } from 'semantic-ui-react';

import { IHit } from '../../../types/search';
import * as s from './hit/Hit.css';

const Hit: React.FunctionComponent<{ data: IHit }> = ({ data }) => (
    <Item>
        <Item.Content>
            <Item.Header
                as={data.url || data.story_url ? 'a' : null}
                href={data.url || data.story_url}
            >
                <p
                    dangerouslySetInnerHTML={{ // Assume the response is sanitised
                        __html: (data._highlightResult.title
                                && data._highlightResult.title.value)
                            || (data._highlightResult.story_title
                                && data._highlightResult.story_title.value)
                            || 'Title not found',
                    }}
                />
            </Item.Header>
            <MediaQuery maxWidth={768}>
                {(matches) => {
                    const size = matches ? 'mini' : 'medium';
                    return (
                        <Item.Meta>
                            <Label className={s.marginTop} size={size}>
                                {data.points || 0} points
                            </Label>
                            <Label className={s.marginTop} size={size}>
                                <Icon name="user" />
                                <span
                                    dangerouslySetInnerHTML={{
                                        __html: data._highlightResult.author
                                            && data._highlightResult.author.value,
                                    }}
                                />
                            </Label>
                            <Label className={s.marginTop} size={size}>
                                {moment(data.created_at).fromNow()}
                            </Label>
                            <Label className={s.marginTop} size={size}>
                                {data.num_comments ? `${data.num_comments} comments` : 'comment'}
                            </Label>
                            {
                                data.url && !matches
                                    ? (
                                        <Label
                                            as="a"
                                            className={s.marginTop}
                                            href={data.url}
                                            size={size}
                                        >
                                            <span
                                                dangerouslySetInnerHTML={{
                                                    __html: data._highlightResult.url
                                                        && data._highlightResult.url.value,
                                                }}
                                            />
                                        </Label>
                                    )
                                    : null
                            }
                        </Item.Meta>
                    );
                }}
            </MediaQuery>
            <Item.Description>
                {
                    data._highlightResult.comment_text || data._highlightResult.story_text
                        ? (
                            <div
                                dangerouslySetInnerHTML={{
                                    __html: (data._highlightResult.comment_text
                                            && data._highlightResult.comment_text.value)
                                        || (data._highlightResult.story_text
                                            && data._highlightResult.story_text.value),
                                }}
                            />
                        )
                        : null
                }
            </Item.Description>
        </Item.Content>
    </Item>
);

export default Hit;
