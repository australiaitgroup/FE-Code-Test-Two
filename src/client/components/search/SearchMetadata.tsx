import * as React from 'react';
import MediaQuery from 'react-responsive';
import { List, Message } from 'semantic-ui-react';

import { ISearchMetadataProps } from '../interfaces/searchMetadata.interface';
import { DateRange } from '../../../types/search';

const SearchMetadata: React.FunctionComponent<ISearchMetadataProps> = ({ nbHits, settings }) => (
    nbHits !== undefined
        ? (
            <MediaQuery minWidth={768}>
                {(matches) => {
                    const [floated, horizontal, size] = matches
                        ? ['right', true, 'small']
                        : [undefined, undefined, 'tiny'] as any;

                    return (
                        <Message info size={size}>
                            <span>{nbHits === 0 ? 'No' : nbHits} results found</span>
                            <List
                                floated={floated}
                                horizontal={horizontal}
                            >
                                <List.Item>Search Type: {settings.searchType}</List.Item>
                                <List.Item>Sort Order: {settings.sortOrder}</List.Item>
                                <List.Item>
                                    Date Range: {
                                    settings.dateRange !== DateRange.Custom
                                        ? settings.dateRange
                                        : `${settings.from} To ${settings.to}`
                                }
                                </List.Item>
                                <List.Item>Hits Per Page: {settings.hitsPerPage}</List.Item>
                            </List>
                        </Message>
                    );
                }}
            </MediaQuery>
        )
        : null
);

export default SearchMetadata;
