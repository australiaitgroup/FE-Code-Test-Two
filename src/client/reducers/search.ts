import {
    HANDLE_INPUT_CHANGE,
    HANDLE_PAGE_CHANGE,
    HANDLE_SEARCH_RESULT,
    SEARCH,
    SET_SEARCH_ERROR,
} from '../actions/search';
import { ISearchResult } from '../../types/search';
import { IAction, ISearchState } from '../../types/stateAndAction';

const initialState = {
    error: null,
    isFetching: false,
    input: '',
    page: 1,
    result: {} as ISearchResult,
    searchFor: '',
};

export default function reducer(
    state: ISearchState = initialState,
    action: IAction,
): ISearchState {
    switch (action.type) {
        case SEARCH:
            return {
                ...state,
                error: null,
                isFetching: true,
                page: 1,
                result: {} as ISearchResult,
                searchFor: action.searchFor,
            };

        case HANDLE_SEARCH_RESULT:
            if (state.searchFor !== action.result.query) return state;
            return { ...state, isFetching: false, result: action.result };

        case HANDLE_INPUT_CHANGE:
            return { ...state, input: action.input };

        case HANDLE_PAGE_CHANGE:
            return { ...state, isFetching: true, page: action.page };

        case SET_SEARCH_ERROR:
            return { ...state, error: action.error, isFetching: false };

        default:
            return state;
    }
}
