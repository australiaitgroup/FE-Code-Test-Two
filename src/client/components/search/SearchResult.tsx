import * as React from 'react';
import { Header, Icon, Item, Message, Segment } from 'semantic-ui-react';

import { ISearchResultProps } from '../interfaces/searchResult.interface';
import Placeholder from '../Placeholder';
import Hit from './Hit';

const SearchResult: React.FunctionComponent<ISearchResultProps> = ({
    error,
    hits,
    isFetching,
    nbHits,
}) => {
    const noMatch = !isFetching && nbHits === 0;

    const getContent = () => {
        if (isFetching) return <Placeholder />;

        if (error) {
            return (
                <Message negative>
                    <Message.Header>Sorry, there was an error</Message.Header>
                    <p>{error.message}</p>
                </Message>
            );
        }

        if (noMatch) {
            return (
                <Header icon>
                    <Icon name="search" />
                    No results matching your query.
                </Header>
            );
        }

        return (
            <Item.Group divided>
                {(hits || []).map(hit => <Hit key={hit.objectID} data={hit} />)}
            </Item.Group>
        );
    };

    return (
        <Segment loading={isFetching} placeholder={noMatch}>
            {getContent()}
        </Segment>
    );
};

export default SearchResult;
