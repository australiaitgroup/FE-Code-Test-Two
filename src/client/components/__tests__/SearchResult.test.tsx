import * as React from 'react';

import SearchResult from '../search/SearchResult';
import { ISearchResult } from '../../../types/search';
import { renderComponent } from '../../../../tools/testHelpers/componentHelpers';

const searchResult = require('../../../../tools/testHelpers/searchTestsMockData.json').searchResult as ISearchResult;

describe.each([
    [
        {
            error: null,
            hits: searchResult.hits,
            isFetching: false,
            nbHits: 1,
        },
        'normal',
    ],
    [
        {
            error: null,
            isFetching: true,
        },
        'loading',
    ],
    [
        {
            error: null,
            hits: [],
            isFetching: false,
            nbHits: 0,
        },
        'no match',
    ],
    [
        {
            error: new Error('message'),
        },
        'error',
    ],
])(
    'SearchResult renders correctly',
    (props, name) => {
        it(`(${name})`, () => {
            const component = renderComponent(<SearchResult {...props} />, {});
            const result = component.toJSON();
            expect(result).toMatchSnapshot();
        });
    },
);
