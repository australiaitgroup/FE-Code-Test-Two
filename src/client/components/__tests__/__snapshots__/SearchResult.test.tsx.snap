// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResult renders correctly (error) 1`] = `
<div
  className="ui segment"
>
  <div
    className="ui negative message"
  >
    <div
      className="header"
    >
      Sorry, there was an error
    </div>
    <p>
      message
    </p>
  </div>
</div>
`;

exports[`SearchResult renders correctly (loading) 1`] = `
<div
  className="ui loading segment"
>
  <div
    className="ui placeholder"
  >
    <div
      className="header"
    >
      <div
        className="line"
      />
      <div
        className="line"
      />
    </div>
    <div
      className="paragraph"
    >
      <div
        className="line"
      />
      <div
        className="line"
      />
      <div
        className="line"
      />
      <div
        className="line"
      />
    </div>
  </div>
</div>
`;

exports[`SearchResult renders correctly (no match) 1`] = `
<div
  className="ui placeholder segment"
>
  <div
    className="ui icon header"
  >
    <i
      aria-hidden="true"
      className="search icon"
    />
    No results matching your query.
  </div>
</div>
`;

exports[`SearchResult renders correctly (normal) 1`] = `
<div
  className="ui segment"
>
  <div
    className="ui divided items"
  >
    <div
      className="item"
    >
      <div
        className="content"
      >
        <a
          className="header"
          href="http://www.npr.org/sections"
        >
          <p
            dangerouslySetInnerHTML={
              Object {
                "__html": "How Well Can You Hear Audio Quality?",
              }
            }
          />
        </a>
        <div
          className="meta"
        >
          <div
            className="ui medium label marginTop"
            onClick={[Function]}
          >
            0
             points
          </div>
          <div
            className="ui medium label marginTop"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="user icon"
            />
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "pingec",
                }
              }
            />
          </div>
          <div
            className="ui medium label marginTop"
            onClick={[Function]}
          >
            3 years ago
          </div>
          <div
            className="ui medium label marginTop"
            onClick={[Function]}
          >
            comment
          </div>
        </div>
        <div
          className="description"
        >
          <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "Similar test: <a href='http://test.tidalhifi.com'",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
